### PR TITLE
Adding validation for email-id field

### LINKED ISSUE #730: Error sending email with capitalization to Apigee.
## Description
Our organization is using SSO to log users into the developer portal. This creates a user in Drupal, and our identity provider is sending over the users email address. The email address gets saved with proper capitalization, so I am `David.Pagini@organization.com`. When the module takes this user and create them as a developer in Apigee, we are getting an error that email address must be lower case.

## Expected Behavior
I would expect that a Developer get created on the Apigee side.

## Screenshots
image

## Notes
I guess I would suggest that all users get set/synced to apigee with a lowercase email address.

## Version Info
Drupal core 9.4.5
Apigee_edge 2.0.4
DevPortal_Kickstart 2.1.0

### PR DESCRIPTION
**Pull request by kedarkhaire**
Fixes #730 
For Hybrid instance (Org type = ApigeeX), 
add field validation for user email field on all operations. Validation will be to check the string is in lowercase. 
Should be triggered on new user creation, user edit, login, SAML auth, drupal default login & LDAP login.

### CODE REVIEW

File: apigee_edge.module [hunk 1]
```diff
@@ -718,6 +718,8 @@ function apigee_edge_form_user_form_alter(&$form, FormStateInterface $form_state
   // connection validation apigee_edge_module_implements_alter() ensures that
   // form_alter hook is called in the last time.
   $validation_functions[] = 'apigee_edge_form_user_form_developer_email_validate';
+  // Validates whether the provided email address is in lowercase.
+  $validation_functions[] = 'apigee_edge_email_field_validate';
```
Comment (mxr576) [inline apigee_edge.module]: Form validation can be bypassed when a user is created via JSONAPI. Entity or field level constraints are the only solutions that all components should obey.

Comment (kedarkhaire) [reply apigee_edge.module]: Hi @mxr576 Sure, I will add field constraints for email field.

File: src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      // First check if the organization is ApigeeX.
+      try {
+        /** @var \Drupal\apigee_edge\SDKConnectorInterface $sdk_connector */
+        $sdk_connector = \Drupal::service('apigee_edge.sdk_connector');
+        $org_controller = \Drupal::service('apigee_edge.controller.organization');
+        /** @var \Apigee\Edge\Api\Management\Entity\Organization $organization */
+        $organization = $org_controller->load($sdk_connector->getOrganization());
+
+        // Check if organization is ApigeeX.
+        if ($organization && ('CLOUD' === $organization->getRuntimeType() || 'HYBRID' === $organization->getRuntimeType())) {
```
Comment (giteshk) [inline src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]: Can this check for ApigeeX instance be moved to the module file instead of here?

File: src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate($value, Constraint $constraint) {
+    foreach ($value as $item) {
+      // First check if the organization is ApigeeX.
+      try {
+        /** @var \Drupal\apigee_edge\SDKConnectorInterface $sdk_connector */
+        $sdk_connector = \Drupal::service('apigee_edge.sdk_connector');
```
Comment (mxr576) [inline src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]: Dependency injection is possible in constraint validators as well. 

See: `\Drupal\Core\Entity\Plugin\Validation\Constraint\ValidReferenceConstraintValidator`

File: src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php (hunk 2 above)
Comment (mxr576) [reply src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]: What would be the benefit of that? The constraint is added in "build-time", however this should be a runtime check because what if a site gets re-connected to an Apigee X org after it was connected to Apigee Edge.

Comment (mxr576) [inline src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]:
```suggestion
        if ($controller->isOrganizationApigeeX()) {
```

https://github.com/apigee/apigee-edge-drupal/blob/0f0e59aff7cc680dd9ec49825728af94df904b84/src/Entity/Controller/OrganizationController.php#L151

This also makes the whole try-catch dead code, if I am not mistaken.

Comment (mxr576) [reply src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]: What may help on potential performance issues caused by fetching org information multiple times (let's say when a batch operation is running) is storing the fact about whether this is an Apigee X org in an in-memory cache (or adding in-memory cache to the org controller if it does not have it yet).

File: src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php (hunk 3 above)
Comment (kedarkhaire) [reply src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]: Updated the code. Thanks for the suggestion.

File: src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php (hunk 2 above)
Comment (kedarkhaire) [reply src/Plugin/Validation/Constraint/LowercaseEmailConstraintValidator.php]: Updated the code. Thanks for the suggestion.

File: src/CliService.php [hunk 4]
```diff
@@ -81,7 +81,7 @@ public function createEdgeRoleForDrupal(
     string $password,
     ?string $base_url,
     ?string $role_name,
-    ?bool $force
+    ?bool $force,
```
Comment (mxr576) [inline src/CliService.php]: IMO these changes should go to a dedicated PR that gets merged independently and before this one.

File: apigee_edge.module [hunk 5]
```diff
@@ -336,6 +337,31 @@ function apigee_edge_entity_base_field_info_alter(&$fields, EntityTypeInterface
...
+/**
+ * Implements hook_form_FORM_ID_alter() for install_configure_form().
+ *
+ * Allows the profile to alter the site configuration form.
+ */
+function apigee_edge_form_install_configure_form_alter(&$form, FormStateInterface $form_state) {
+  $form['#validate'][] = 'apigee_edge_checkemail';
+}
+
+function apigee_edge_checkemail(&$form, FormStateInterface $form_state) {
```
Comment (mxr576) [inline apigee_edge.module]:
```suggestion
/**
*  @internal
*/
function _apigee_edge_site_install_form_check_email(array &$form, FormStateInterface $form_state): void {
```

The benefit is marking this as internal with all power, nobody should use this function outside of the module.

But it still deserves a better doc block.

File: apigee_edge.module [hunk 6]
```diff
@@ -336,6 +337,31 @@ function apigee_edge_entity_base_field_info_alter(&$fields, EntityTypeInterface
...
+  try {
+    $org_controller = \Drupal::service('apigee_edge.controller.organization');
+    // Check if org is ApigeeX.
+    if ($org_controller->isOrganizationApigeeX()) {
```
Comment (mxr576) [inline apigee_edge.module]: If the code logic from LowercaseEmailConstraintValidator would be decoupled to a dedicated service then the code could be more DRY because the validation logic would be centralized into one source of truth that both code could leverage.

File: apigee_edge.module (hunk 5 above)
Comment (kedarkhaire) [reply apigee_edge.module]: Noted, will apply the change

File: src/CliService.php (hunk 4 above)
Comment (kedarkhaire) [reply src/CliService.php]: Updated. moved to #1051

File: apigee_edge.module (hunk 5 above)
Comment (kedarkhaire) [reply apigee_edge.module]: Updated

**Review by shishir-intelli — APPROVED** [review]
LGTM
